postreise/plot/plot_carbon_map.py

```diff
@@ -1,533 +1,319 @@
+    :param float scale_factor: scaling factor for size of emissions circles glyphs.
+    :param bool background_map: whether to plot map tiles behind figure.
+    :return: (*bokeh.plotting.figure*) -- carbon emission map.
+    """
     p = figure(
-        tools=tools,
+        tools="pan,wheel_zoom,reset,save",
         x_axis_location=None,
         y_axis_location=None,
-        plot_width=800,
-        plot_height=800,
+        plot_width=figsize[0],
+        plot_height=figsize[1],
         output_backend="webgl",
-        sizing_mode="stretch_both",
+        sizing_mode="scale_both",
         match_aspect=True,
     )
-    p_legend = figure(
-        x_axis_location=None,
-        y_axis_location=None,
-        toolbar_location=None,
-        plot_width=200,
-        plot_height=400,
-        y_range=(0, 4),
-        x_range=(0, 2),
-        output_backend="webgl",
-    )
-
-    # circle glyphs that exist only for the web legend,
-    # these are plotted behind the tiles, not visible
-    p.circle(
-        -8.1e6,
-        5.2e6,
-        fill_color=color_coal,
-        color=color_coal,
-        alpha=0.5,
-        size=50,
-        legend_label=label_coal,
-    )
-    p.circle(
-        -8.1e6,
-        5.2e6,
-        fill_color=color_ng,
-        color=color_ng,
-        alpha=0.5,
-        size=50,
-        legend_label=label_ng,
-    )
-
-    p.add_tile(get_provider(Vendors.CARTODBPOSITRON_RETINA))
-    # state borders
-    p.patches(a, b, fill_alpha=0.0, line_color="gray", line_width=2)
-
-    # emissions circles, web version
-    if web:
-
-        circle = p.circle(
-            "x",
-            "y",
-            fill_color="color",
-            color="color",
-            alpha=0.25,
-            radius="radius",
-            source=bus_source,
-        )
-
-    else:
-        p.circle(
-            "x",
-            "y",
-            fill_color="color",
-            color="color",
-            alpha=0.25,
-            size="size",
-            source=bus_source,
-        )
-
-    # legend: white square background and bars per color code
-    p_legend.square(1, [1, 3], fill_color="white", color="white", size=300)
-    p_legend.square(
-        1, [1, 3], fill_color="white", color="black", size=(2000000 / 100) ** 0.5
-    )
-    p_legend.circle(
-        1,
-        y=np.repeat([1, 3], 3),
-        fill_color=np.repeat([color_coal, color_ng], 3),
-        color=np.repeat([color_coal, color_ng], 3),
+    p.xgrid.visible = False
+    p.ygrid.visible = False
+
+    plot_states(bokeh_figure=p, background_map=background_map)
+
+    # emissions as circles
+    source = {
+        "x": emission_map["x"],
+        "y": emission_map["y"],
+        "size": np.log2(emission_map["amount"]) * scale_factor,
+        "tons": emission_map["amount"],
+        "color": emission_map["color"],
+        "type": emission_map["type"],
+    }
+    circle = p.circle(
+        "x",
+        "y",
+        fill_color="color",
+        color="color",
```

Comment:
If we specify `color`, we don't need to also specify `fill_color`: `color` sets everything all at once.
https://docs.bokeh.org/en/latest/docs/reference/plotting.html#bokeh.plotting.Figure.circle

Reply:
Done